commit/plugin.go

```diff
@@ -85,8 +102,147 @@ func NewPlugin(
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (p *Plugin) Observation(
+	ctx context.Context, outCtx ocr3types.OutcomeContext, _ types.Query,
+) (types.Observation, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+	fChain := p.ObserveFChain()
+	//TODO: Send fchain instead of computing it inside
+	merkleObs, err := p.merkleProcessor.Observation(ctx, prevOutcome.MerkleRootOutcome, merkleroot.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get merkle observation", "err", err)
+	}
+	tokenPricesObs, err := p.tokenProcessor.Observation(ctx, prevOutcome.TokenPriceOutcome, tokenprice.Query{})
+	if err != nil {
+		//log error
+		p.lggr.Errorw("failed to get token prices", "err", err)
+	}
+	gasObs, err := p.feeProcessor.Observation(ctx, prevOutcome.ChainFeeOutcome, chainfee.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get gas prices", "err", err)
 	}
+
+	obs := Observation{
+		MerkleObs: merkleObs,
+		TokenObs:  tokenPricesObs,
+		GasObs:    gasObs,
+		FChain:    fChain,
+	}
+	return obs.Encode()
+}
+
+func (p *Plugin) ObserveFChain() map[cciptypes.ChainSelector]int {
+	fChain, err := p.homeChain.GetFChain()
+	if err != nil {
+		// TODO: metrics
+		p.lggr.Warnw("call to GetFChain failed", "err", err)
+		return map[cciptypes.ChainSelector]int{}
+	}
+	return fChain
+}
+
+// Outcome depending on the current state, either:
+// - chooses the seq num ranges for the next round
+// - builds a report
+// - checks for the transmission of a previous report
+func (p *Plugin) Outcome(
+	outCtx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation,
+) (ocr3types.Outcome, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+
+	var merkleObservations []MerkleRootObservation
+	var tokensObservations []TokenPricesObservation
+	var feeObservations []ChainFeeObservation
+
+	for _, ao := range aos {
+		obs, err := DecodeCommitPluginObservation(ao.Observation)
+		if err != nil {
+			p.lggr.Errorw("failed to decode observation", "err", err)
+			continue
+		}
+		merkleObservations = append(merkleObservations,
+			MerkleRootObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.MerkleObs,
+			},
+		)
+
+		tokensObservations = append(tokensObservations,
+			TokenPricesObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.TokenObs,
+			},
+		)
+
+		feeObservations = append(feeObservations,
+			ChainFeeObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.GasObs,
+			},
+		)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(3)
+
+	var merkleOutcome merkleroot.Outcome
+	var tokensOutcome tokenprice.Outcome
+	var gasOutcome chainfee.Outcome
+
+	go func(mOut *merkleroot.Outcome) {
+		defer wg.Done()
+		outcome, err := p.merkleProcessor.Outcome(
+			prevOutcome.MerkleRootOutcome,
+			merkleroot.Query{},
+			merkleObservations,
+		)
+		if err != nil {
+			p.lggr.Errorw("failed to get merkle outcome", "err", err)
+		}
+
+		*mOut = outcome
+	}(&merkleOutcome)
```

Comment:
Although this is technically safe because the pointer is only accessed by a single goroutine its unidiomatic and I don't really like it.

Reply:
Same for other goroutines